libfwupdplugin/fu-struct.c

```diff
@@ -0,0 +1,1465 @@
+/*
+ * Copyright (C) 2023 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuStruct"
+
+#include "config.h"
+
+#include <gio/gio.h>
+#include <string.h>
+
+#include "fu-byte-array.h"
+#include "fu-common.h"
+#include "fu-mem.h"
+#include "fu-string.h"
+#include "fu-struct.h"
+
+/**
+ * FuStruct:
+ *
+ * A smart structure that supports endian conversion, arrays, GUIDs, strings, default and constant
+ * data of variable size.
+ *
+ * In most cases the smart structure will be defined in the #GObject `_init()` function:
+ *
+ *    fu_struct_register(self,
+ *                       "StructureName {"
+ *                       "    member_id: u16le,"
+ *                       "}");
```

Comment:
I don't really like the defining how it looks in a big string and parsing that.  Can we not instead just have a series of calls that define the name of the structure and the endiannness and type of each member?

Reply:
I actually tried that; it gets very verbose. You need to essentially have FuStructItem exported, and then have getters and setters for the id, the type, the endian, the default value, the flags to indicate it's constant. It's at least 3 lines per struct field. Plus, I'd quite like to keep them in Rust format so that when we do start using it in core code we can just use the struct without big changes.

Even one line it's `fu_struct_add_member(st, "member_id", FU_STRUCT_FMT_U16, G_LITTLE_ENDIAN, NULL, FU_STRUCT_FLAG_NONE)` which gets split by clang-format into 6 lines....